SpatialGDK/Source/SpatialGDK/Private/SpatialView/EntityComponentRecord.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialView/EntityComponentRecord.h"
+
+namespace SpatialGDK
+{
+
+void EntityComponentRecord::AddComponent(Worker_EntityId EntityId, ComponentData Data)
+{
+	const EntityComponentId Id = { EntityId, Data.GetComponentId() };
+
+	// If the component is recorded as removed then transition to complete-updated.
+	// otherwise record it as added.
+	if (ComponentsRemoved.RemoveSingleSwap(Id))
+	{
+		UpdateRecord.AddComponentDataAsUpdate(EntityId, MoveTemp(Data));
+	}
+	else
+	{
+		ComponentsAdded.Push(EntityComponentData{ EntityId, MoveTemp(Data) });
+	}
+}
+
+void EntityComponentRecord::RemoveComponent(Worker_EntityId EntityId, Worker_ComponentId ComponentId)
+{
+	const EntityComponentId Id = { EntityId, ComponentId };
+	const EntityComponentData* FoundComponentAdded = ComponentsAdded.FindByPredicate(EntityComponentIdEquality{ Id });
+
+	// If the component is recorded as added then erase the record.
+	// Otherwise record it as removed (additionally making sure it isn't recorded as updated).
+	if (FoundComponentAdded)
+	{
+		ComponentsAdded.RemoveAtSwap(FoundComponentAdded - ComponentsAdded.GetData());
+	}
+	else
+	{
+		UpdateRecord.RemoveComponent(EntityId, ComponentId);
+		ComponentsRemoved.Push(Id);
+	}
+}
+
+void EntityComponentRecord::AddComponentAsUpdate(Worker_EntityId EntityId, ComponentData Data)
```

Comment:
What is the difference between AddComponent and AddComponentAsUpdate ? Also, why are we not checking the Removed array ? I assume we receive a consistent stream, where we can't have a add->remove->update stream of ops.

Reply:
AddAsUpdate is because we can get Add -> Update -> Add. And that second add takes the form of a new component but is secretly an update.
And yes you know at the point this is called that it can't be in the remove list. It can only recorded as added or updated at this point.

Reply:
The way you know it's an Add-As-Update instead of a regular add is because the view registered that it had already seen this component.